src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+  job: IngestJobV2<S, Opr>;
+  createRequest: CreateIngestJobV2Request;
+};
+
+type CreateQueryJobV2Options<S extends Schema> = {
+  connection: Connection<S>;
+  pollingOptions: BulkV2PollingOptions;
+  body: QueryJobBody;
+};
+
+type BulkV2PollingOptions = {
+  pollInterval: number;
+  pollTimeout: number;
+};
+
+class JobPollingTimeoutError extends Error {
+  jobId: string;
+
+  /**
+   *
+   */
+  constructor(message: string, jobId: string) {
+    super(message);
+    this.name = 'JobPollingTimeout';
+    this.jobId = jobId;
+  }
+}
+
+class BulkApiV2<S extends Schema> extends HttpApi<S> {
+  hasErrorInResponseBody(body: any) {
+    return (
+      Array.isArray(body) &&
+      typeof body[0] === 'object' &&
+      'errorCode' in body[0]
+    );
+  }
+
+  isSessionExpired(response: HttpResponse): boolean {
+    return (
+      response.statusCode === 401 && /INVALID_SESSION_ID/.test(response.body)
+    );
+  }
+
+  parseError(body: any) {
+    return {
+      errorCode: body[0].errorCode,
+      message: body[0].message,
+    };
+  }
+}
+
+export class BulkV2<S extends Schema> {
+  private connection: Connection<S>;
+  private logger: Logger;
+
+  /**
+   * Polling interval in milliseconds
+   */
+  pollInterval = 1000;
+
+  /**
+   * Polling timeout in milliseconds
+   */
+  pollTimeout = 10000;
+
+  constructor(connection: Connection<S>) {
+    this.connection = connection;
+    this.logger = connection._logger;
+    this.logger = this.connection._logLevel
+      ? getLogger('bulk2').createInstance(this.connection._logLevel)
+      : getLogger('bulk2');
+  }
+
+  /**
+   * Create an instance of an ingest job object.
+   *
+   * @params {NewIngestJobOptions} options object
+   * @returns {IngestJobV2} An ingest job instance
+   * @example
+   * // Upsert records to the Account object.
+   *
+   * const job = connection.bulk2.createJob({
+   *   operation: 'insert'
+   *   object: 'Account',
+   * });
+   *
+   * // create the job in the org
+   * await job.open()
+   *
+   * // upload data
+   * await job.uploadData(csvFile)
+   *
+   * // finished uploading data, mark it as ready for processing
+   * await job.close()
+   */
+  public createJob<Opr extends IngestOperation>(
+    options: NewIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Get an ingest job instance specified by a given job ID
+   *
+   * @param options Options object with a job ID
+   * @returns IngestJobV2 An ingest job
+   */
+  public job<Opr extends IngestOperation>(
+    options: ExistingIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Create, upload, and start bulkload job
+   */
+  async loadAndWaitForResults(
```

Comment:
this name was chosen over `load` here:
https://salesforce.quip.com/CcboAPmPbgYb#TZAADAy1B6x
https://github.com/jsforce/jsforce/pull/1251#pullrequestreview-1062415091

await, promise gets resolved:
![Screenshot 2024-02-14 at 15 29 10](https://github.com/jsforce/jsforce/assets/6853656/d423fc12-f48d-4614-9357-52d2c5d065ec)

if you don't await you get the Batch (writable stream):
![Screenshot 2024-02-14 at 15 29 52](https://github.com/jsforce/jsforce/assets/6853656/8b1035b3-16fe-474a-a035-b703334e6094)

which you can write csv into (record stream system, there's an internal csv converter):
![Screenshot 2024-02-14 at 15 45 03](https://github.com/jsforce/jsforce/assets/6853656/b1ddd963-90c6-4af4-988d-4b2edb7d2b87)

Reply:
a few questions:

1) do we still want to call it `loadAndWaitForResults`? I would be fine calling it `load`, even if it doesn't offer the same API as `conn.bulk.load`.

2) should it support returning a writable stream like `conn.bulk` that you can pipe csv into?
`loadAndWaitForResults` already accepts a readable input.

I didn't implement it because bulk2 requires the job to be closed for the data to be processed, so it would require some event handling to mark the job as "upload complete" once the data is written.

Reply:
> Does Readable the readable input allow the readable side of a duplex/pipe? Like, obviously fs.createReadStream could be input here from a local file. What about a transform stream or pipe where someone wants to

yes, you could use a transform stream as long as you transform it back to plain csv.
https://github.com/jsforce/jsforce/blob/3.0/src/record-stream.ts#L235

> What's the use case for writeable?

consumers can start writing data early instead of having `loadAndWaitForResults` iterate over the array and write them individually.

here's an example (you can choose which Bulk API to use in some CRUD methods, this is how data is handled internally): https://github.com/jsforce/jsforce/blob/e760c2843d8d23fd57331661699dc538784f7a1c/src/query.ts#L950-L953